File: leetcode/1964.找出到每个位置为止最长的有效障碍赛跑路线.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& nums) {
        //lcs问题的变种
        int n = nums.size();
        vector<int> q;
        vector<int> ans;
        for(int i = 0; i < n; i++){
            int l = 0, r = q.size() - 1;
            while(l < r){
                int mid = (l + r) / 2;
                if(q[mid] > nums[i]){
                    r = mid;
                }
                else{
                    l = mid + 1;
                }
            }
            if(q.empty() || q[l] <= nums[i]){
                q.push_back(nums[i]);
                ans.push_back(q.size());
            }
            else{
                q[l] = nums[i];
                ans.push_back(l+1);
            }
        }
        return ans;
    }
};

File: leetcode/1964.找出到每个位置为止最长的有效障碍赛跑路线.py
class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        n = len(obstacles)
        q = []
        ans = []
        for x in obstacles:
            l, r = 0, len(q) - 1
            while l < r:
                mid = (l + r) // 2
                if q[mid] > x:
                    r = mid
                else:
                    l = mid + 1
            if not q or q[l] <= x:
                q.append(x)
                l = len(q)
            else:
                q[l] = x
                l += 1
            ans.append(l)
        return ans
